root/programs/C#/Frameworks/Infrastructure/ServiceInterface/AsyncProcessingService/AsyncProcessingService.cs

```diff
@@ -0,0 +1,222 @@
+                    _asyncParameterValue.CommandId = 0;
+                    break;
+            }
+            _asyncParameterValue.NumberOfRetries = numberOfRetries;
+            _asyncParameterValue.CompletionDateTime = completionDateTime;
+
+            return _asyncParameterValue;
+        }
+
+        /// <summary>
+        /// Set things in motion so your service can do its work.
+        /// </summary>
+        protected override void OnStart(string[] args)
+        {
+            StartByThreadPool(AsyncProcessingServiceParameterValue.AsyncStatus.Processing);
+        }
+
+        /// <summary>
+        /// Calls callback method using Threadpool worker item.
+        /// </summary>
+        /// <param name="check"></param>
+        /// <returns></returns>
+        public bool StartByThreadPool(AsyncProcessingServiceParameterValue.AsyncStatus check)
+        {
+            int maxThreadCount;
+
+            lock (this)
+            {
+                maxThreadCount
+                    = int.Parse(ConfigurationManager.AppSettings.Get("FxMaxThreadCount").ToString());
+
+                //Checks if threadcount greater than maxthreadcount sets in config file 
+                //then service status will be inserted or updated as Abort
+                if (_threadCount >= maxThreadCount)
+                {
+                    UpdateServiceData(AsyncProcessingServiceParameterValue.AsyncStatus.Abort, "A", DateTime.Now, DateTime.Now, _threadCount, 0, DateTime.Now, 0, 0);
+                    return false;
+                }
+
+                _threadCount++;
+
+                _asyncParameterValue = UpdateServiceData(check, "A", DateTime.Now, DateTime.Now, _threadCount, 0, DateTime.Now, 0, 0);
+
+                ThreadPool.QueueUserWorkItem(new WaitCallback(InvokeController), (object)_asyncParameterValue);
+            }
+            return true;
+        }
+
+        /// <summary>
+        /// Updates using LayerB through callcontroller class
+        /// </summary>
+        /// <param name="_asyncParameterValue"></param>
+        private void InvokeController(object _asyncParameterValue)
+        {
+            AsyncProcessingServiceReturnValue asyncReturnValue;
+
+            // InProcess Call using CallController for updating service in database
+            CallController callController = new CallController(this._asyncParameterValue.User);
+            asyncReturnValue = (AsyncProcessingServiceReturnValue)callController.Invoke("UpdateInProcessCall", _asyncParameterValue);
+        }
+
+        /// <summary>
+        /// Stop this service.
+        /// </summary>
+        protected override void OnStop()
+        {
+            StartByThreadPool(AsyncProcessingServiceParameterValue.AsyncStatus.Stop);
+        }
+
+        /// <summary>
+        /// Pause this service.
+        /// </summary>
+        protected override void OnPause()
```

Comment:
- ServiceBase.OnPause Method (System.ServiceProcess)  
https://msdn.microsoft.com/library/system.serviceprocess.servicebase.onpause.aspx
- In this method, I want to pause this windows service.  
Because Pause, only stop the monitoring by the main thread.